libevmasm/ThirdOptimizerTesting.h

```diff
@@ -0,0 +1,177 @@
+		m_ptr->m_instruction = _instruction;
+		m_ptr->m_operands = _operands;
+	}
+
+	NewOptimizerPattern(Instruction const& _instruction) : NewOptimizerPattern(_instruction, {}) {}
+
+	NewOptimizerPattern(AssemblyItemType const& _type) : m_ptr(std::make_shared<Underlying>())
+	{
+		m_ptr->m_kind = Kind::Any;
+		m_ptr->m_assemblyType = _type;
+	}
+
+	NewOptimizerPattern(u256 const& _constant) : m_ptr(std::make_shared<Underlying>())
+	{
+		m_ptr->m_kind = Kind::Constant;
+		m_ptr->m_constant = _constant;
+	}
+
+	NewOptimizerPattern(unsigned const& _constant) : NewOptimizerPattern(u256(_constant)) {}
+
+	NewOptimizerPattern(Kind const& _kind) : m_ptr(std::make_shared<Underlying>()) { m_ptr->m_kind = _kind; }
+
+	Kind kind() const { return m_ptr->m_kind; }
+
+	vector<NewOptimizerPattern> const& operands() const
+	{
+		assertThrow(hasOperationValues(), OptimizerException, "invalid request for operands");
+		return m_ptr->m_operands;
+	}
+
+	vector<NewOptimizerPattern>& operands()
+	{
+		assertThrow(hasOperationValues(), OptimizerException, "invalid request for operands");
+		return m_ptr->m_operands;
+	}
+
+	Instruction instruction() const
+	{
+		assertThrow(hasOperationValues(), OptimizerException, "invalid request for instruction");
+		return m_ptr->m_instruction;
+	}
+
+	u256 constant() const
+	{
+		assertThrow(hasConstant(), OptimizerException, "invalid request for constant");
+		return m_ptr->m_constant.get();
+	}
+
+	AssemblyItemType assemblyItemType() const
+	{
+		assertThrow(hasAssemblyItemType(), OptimizerException, "invalid request for assembly item type");
+		return m_ptr->m_assemblyType.get();
+	}
+
+	bool matches(vector<AssemblyItem> const& _items, bool _unbind = true);
+
+	u256 d() const
+	{
+		if (isBound())
+		{
+			if (kind() == Kind::Constant)
+				return boundItems().back().data();
+			assertThrow(false, OptimizerException, "invalid request for constant");
+		}
+		else
+			return constant();
+	}
+
+	bool isBound() const { return m_ptr->m_boundItems.is_initialized(); }
+
+	vector<AssemblyItem> const& boundItems() const
+	{
+		assertThrow(isBound(), OptimizerException, "invalid request for bound items");
+		return m_ptr->m_boundItems.get();
+	}
+
+	bool hasOperationValues() const { return m_ptr->m_hasOperationValues; }
+	bool hasConstant() const { return m_ptr->m_constant.is_initialized(); }
+	bool hasAssemblyItemType() const { return m_ptr->m_assemblyType.is_initialized(); }
+
+private:
+	void bind(std::vector<AssemblyItem> const& _items);
+	void unbind()
+	{
+		m_ptr->m_boundItems.reset();
+
+		if (hasOperationValues())
+		{
+			for (auto& operand : operands()) operand.unbind();
+		}
+	}
+
+	struct Underlying
+	{
+		boost::optional<std::vector<AssemblyItem>> m_boundItems;
+
+		Kind m_kind;
+
+		boost::optional<u256> m_constant;
+
+		bool m_hasOperationValues = false;
+		Instruction m_instruction;
+		vector<NewOptimizerPattern> m_operands;
+
+		boost::optional<AssemblyItemType> m_assemblyType;
+	};
+
+	std::shared_ptr<Underlying> m_ptr;
```

Comment:
What is the purpose of this?

Reply:
To create reference semantics. Patterns for operation operands have to be copied in to the parent (otherwise we could reference a pattern outside its lifetime). If we were to use plain value-fields, since they are copied in, in RulesList, the pattern that matches the opcodes will have copies of X and Y, which means the copies will be bound, and the replacement lambda captures by value, so the pattern matching (which is the one that is bound) and the replacement (which is what items are created from) are disjoint.

This means that once the pattern is bound in matching, the replacement lambda's version will also be bound.

Reply:
The other implementations solve this via `matchGroups`, but your version might be more elegant. Have you considered using `shared_ptr`s instead? I.e. put all the members directly into the class but reference operands through shared pointers.

Reply:
That would be equivalent, just with more reference counting, so it's up to style. If you want that change, it's fine.

Reply:
Oh, might have misread that. Are you suggesting only operands by under a shared_ptr?

Reply:
I'm suggesting moving the members into the outer class and using explicit shared_ptrs everywhere. The benefit would be that you would not have to use `m_ptr-><member>` all the time and the constructor would also be way simpler.

Reply:
So make every field a direct shared_ptr member instead of having a shared_ptr to a struct?

Reply:
No, just make every use of `NewOptimizerPattern` a `shared_ptr<NewOptimizerPattern>`